src/optimizer/cardinality_estimator.cpp

```diff
@@ -0,0 +1,569 @@
+
+static TableCatalogEntry *GetCatalogTableEntry(LogicalOperator *op) {
+	if (op->type == LogicalOperatorType::LOGICAL_GET) {
+		auto get = (LogicalGet *)op;
+		TableCatalogEntry *entry = get->GetTable();
+		return entry;
+	}
+	for (auto &child : op->children) {
+		TableCatalogEntry *entry = GetCatalogTableEntry(child.get());
+		if (entry != nullptr) {
+			return entry;
+		}
+	}
+	return nullptr;
+}
+
+bool CardinalityEstimator::SingleColumnFilter(FilterInfo *filter_info) {
+	if (filter_info->left_set && filter_info->right_set) {
+		// Both set
+		return false;
+	}
+	// Filter on one relation, (i.e string or range filter on a column).
+	// Grab the first relation and add it to the the equivalence_relations
+	D_ASSERT(filter_info->set->count >= 1);
+	for (const RelationsToTDom &r2tdom : relations_to_tdoms) {
+		auto &i_set = r2tdom.equivalent_relations;
+		if (i_set.count(filter_info->left_binding) > 0) {
+			// found an equivalent filter
+			return true;
+		}
+	}
+	auto key = ColumnBinding(filter_info->left_binding.table_index, filter_info->left_binding.column_index);
+	column_binding_set_t tmp({key});
+	relations_to_tdoms.emplace_back(RelationsToTDom(tmp));
+	return true;
+}
+
+vector<idx_t> CardinalityEstimator::DetermineMatchingEquivalentSets(FilterInfo *filter_info) {
+	vector<idx_t> matching_equivalent_sets;
+	auto equivalent_relation_index = 0;
+	// eri = equivalent relation index
+	bool added_to_eri;
+
+	for (const RelationsToTDom &r2tdom : relations_to_tdoms) {
+		auto &i_set = r2tdom.equivalent_relations;
+		added_to_eri = false;
+		if (i_set.count(filter_info->left_binding) > 0) {
+			matching_equivalent_sets.push_back(equivalent_relation_index);
+			added_to_eri = true;
+		}
+		// don't add both left and right to the matching_equivalent_sets
+		// since both left and right get added to that index anyway.
+		if (i_set.count(filter_info->right_binding) > 0 && !added_to_eri) {
+			matching_equivalent_sets.push_back(equivalent_relation_index);
+		}
+		equivalent_relation_index += 1;
+	}
+	return matching_equivalent_sets;
+}
+
+void CardinalityEstimator::AddToEquivalenceSets(FilterInfo *filter_info, vector<idx_t> matching_equivalent_sets) {
+	D_ASSERT(matching_equivalent_sets.size() <= 2);
+	if (matching_equivalent_sets.size() > 1) {
+		// an equivalence relation is connecting to sets of equivalence relations
+		// so push all relations from the second set into the first. Later we will delete
+		// the second set.
+		for (ColumnBinding i : relations_to_tdoms.at(matching_equivalent_sets[1]).equivalent_relations) {
+			relations_to_tdoms.at(matching_equivalent_sets[0]).equivalent_relations.insert(i);
+		}
+		relations_to_tdoms.at(matching_equivalent_sets[1]).equivalent_relations.clear();
+		relations_to_tdoms.at(matching_equivalent_sets[0]).filters.push_back(filter_info);
+		// add all values of one set to the other, delete the empty one
+	} else if (matching_equivalent_sets.size() == 1) {
+		idx_t set_i = matching_equivalent_sets.at(0);
```

Comment:
Nitpick: This code would have a lot less repetition if you get a reference to the element from the map first, e.g.:
```c++
auto &tdom_i = relations_to_tdoms(matching_equivalent_sets.at(0));
tdom_i.equivalent_relations.insert(filter_info->left_binding);
tdom_i.equivalent_relations.insert(filter_info->right_binding);
tdom_i.filters.push_back(filter_info);
```